File: problems/heaps/kth_largest_element_in_array.py
"""
LeetCode 215. Kth Largest Element in an Array

Return the kth largest element (not the kth distinct one) in an unsorted
array.

Example:
    [3,2,1,5,6,4], k = 2 -> 5

Approach: maintain a min-heap of size k as we scan the array. After
processing all elements, the heap's smallest element (its root) is the
kth largest overall.

Time:  O(n log k)
Space: O(k)
"""

import heapq
from typing import List


def find_kth_largest(nums: List[int], k: int) -> int:
    heap = nums[:k]
    heapq.heapify(heap)
    for num in nums[k:]:
        if num > heap[0]:
            heapq.heapreplace(heap, num)
    return heap[0]


if __name__ == "__main__":
    assert find_kth_largest([3, 2, 1, 5, 6, 4], 2) == 5
    assert find_kth_largest([3, 2, 3, 1, 2, 4, 5, 5, 6], 4) == 4
    assert find_kth_largest([1], 1) == 1
    print("All tests passed.")

File: solutions/problems/heaps/kthLargestElementInArray.ts
/**
 * Approach: maintain a min-heap of size k as we scan the array. After
 * processing all elements, the heap's smallest element (its root) is the
 * kth largest overall. Implemented here with a small array-backed heap.
 * Time: O(n log k). Space: O(k).
 */

export function findKthLargest(nums: number[], k: number): number {
  const heap: number[] = [];

  const siftUp = (i: number) => {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (heap[i] < heap[parent]) {
        [heap[i], heap[parent]] = [heap[parent], heap[i]];
        i = parent;
      } else break;
    }
  };
  const siftDown = (i: number) => {
    const n = heap.length;
    while (true) {
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      let smallest = i;
      if (l < n && heap[l] < heap[smallest]) smallest = l;
      if (r < n && heap[r] < heap[smallest]) smallest = r;
      if (smallest === i) break;
      [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
      i = smallest;
    }
  };

  for (const num of nums) {
    if (heap.length < k) {
      heap.push(num);
      siftUp(heap.length - 1);
    } else if (num > heap[0]) {
      heap[0] = num;
      siftDown(0);
    }
  }
  return heap[0];
}
